clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -195,11 +199,23 @@ public boolean completeExceptionally(Throwable newException) {
             waiters = null;
         }
         for (BiConsumer<? super T, ? super Throwable> waiter : oldWaiters) {
-            waiter.accept(null, newException);
+            invokeWaiter(waiter, null, newException);
         }
         return true;
     }
 
+    private void invokeWaiter(
+        BiConsumer<? super T, ? super Throwable> waiter,
+        T value,
+        Throwable exception
+    ) {
+        try {
+            waiter.accept(value, exception);
+        } catch (Exception e) {
+            log.error("Exception when invoking action on future completion {}", waiter, e);
```

Comment:
fyi: are you aware CompletableFuture does not swallow the exception, but propagates it?

`addWaiter` is there called[ `whenComplete()`](https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/CompletableFuture.html#whenComplete-java.util.function.BiConsumer-) and it has the following behavior:

>  so the supplied action should not throw an exception. However, if it does, the following rules apply: if this stage completed normally but the supplied action throws an exception, then the returned stage completes exceptionally with the supplied action's exception. 

I am not saying you have to implement it like this.  As I am aware this behaviour is a lot more difficult to implement and it changes the public api of addWaiter.  I just want to make sure you know.

Reply:
oops, meant to reply here, see my answer in the main thread below.